PayPalNative/src/main/java/com/braintreepayments/api/PayPalNativeClient.java

```diff
@@ -50,33 +79,130 @@ public void onResult(Configuration configuration, Exception error) {
         internalPayPalClient.sendRequest(activity, payPalRequest, new PayPalInternalClientCallback() {
             @Override
-            public void onResult(PayPalResponse payPalResponse, Exception error) {
+            public void onResult(final PayPalResponse payPalResponse, final Exception error) {
                 if (payPalResponse != null) {
-                    // TODO: start native checkout
-                    callback.onResult(new BraintreeException("TODO: integrate native checkout with Braintree SDK."));
+                    startNativeCheckout(activity, configuration, payPalClientId, payPalResponse, callback);
                 } else {
-                    callback.onResult(error);
+                    callback.onResult(null, error);
                 }
             }
         });
     }
 
+    private void startNativeCheckout(final FragmentActivity activity, final Configuration configuration, final String clientId, final PayPalResponse payPalResponse, final PayPalNativeFlowCallback callback) {
+        final String redirectUrl = getPayPalReturnUrl(activity);
+        Environment environment = Environment.SANDBOX;
+        if ("production".equalsIgnoreCase(configuration.getEnvironment())) {
+            environment = Environment.LIVE;
+        }
+
+        CheckoutConfig checkoutConfig = new CheckoutConfig(
+                activity.getApplication(),
+                clientId,
+                environment,
+                redirectUrl
+        );
+        PayPalCheckout.setConfig(checkoutConfig);
+        PayPalCheckout.start(
+                new CreateOrder() {
+                    @Override
+                    public void create(final CreateOrderActions createOrderActions) {
+                        createOrderActions.set(payPalResponse.getPairingId());
+                    }
+                },
+                new OnApprove() {
+                    @Override
+                    public void onApprove(final Approval approval) {
+                        braintreeClient.sendAnalyticsEvent("paypal.native.approved");
+                        approved(approval, payPalResponse, callback);
+                    }
+                },
+                new OnCancel() {
+                    @Override
+                    public void onCancel() {
+                        braintreeClient.sendAnalyticsEvent("paypal.native.client_cancel");
+                        callback.onResult(null, new BraintreeException("User Cancelled"));
+                    }
+                },
+                new OnError() {
+                    @Override
+                    public void onError(final ErrorInfo errorInfo) {
+                        braintreeClient.sendAnalyticsEvent("paypal.native.error");
+                        callback.onResult(null, new BraintreeException(errorInfo.getReason()));
```

Comment:
I delegate both `OnCancel` and `OnError` to the callback. I have doubts about `OnCancel` though, as it is technically not an error, but should we add another parameter to `PayPalNativeFlowCallback`?

Reply:
We have documented that a `BraintreeException` with a message indicating user cancellation will be returned on user cancellation for all the payment flows in the SDK, so I think it's good to go with that approach here as well. It also allows us to maintain the two-parameter callback pattern.